File: python/circular-buffer/circular_buffer.py
"""Circular buffer."""


class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full."""


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty."""


class CircularBuffer:
    """Implement a circular buffer."""

    _read: int
    _write: int

    def __init__(self, capacity: int):
        """Initialize a buffer."""
        self._len = capacity + 1
        self.clear()

    def _next(self, pos: int) -> int:
        """Return the next position."""
        return (pos + 1) % self._len

    @property
    def full(self) -> bool:
        """Return if the buffer is full."""
        return self._read == self._next(self._write)

    @property
    def empty(self) -> bool:
        """Return if the buffer is empty."""
        return self._read == self._write

    def read(self) -> str:
        """Return the next value in the buffer."""
        if self.empty:
            raise BufferEmptyException("Circular buffer is empty")
        data = self._data[self._read]
        self._read = self._next(self._read)
        return data

    def write(self, data: str) -> None:
        """Write a value to the buffer."""
        if self.full:
            raise BufferFullException("Circular buffer is full")
        self._data[self._write] = data
        self._write = self._next(self._write)

    def overwrite(self, data: str) -> None:
        """Write a value to the buffer, overwriting when full."""
        if self.full:
            self.read()
        self.write(data)

    def clear(self):
        """Reset the buffer."""
        self._data = [None] * self._len
        self._read = 0
        self._write = 0

File: go/circular-buffer/circular_buffer.go
// Package circular implements a circular buffer.
package circular

import "errors"

// Buffer provides a circular buffer.
type Buffer struct {
	data  []byte
	read  int
	write int
}

// NewBuffer returns a new buffer.
func NewBuffer(size int) *Buffer {
	return &Buffer{data: make([]byte, size + 1)}
}

func (b *Buffer) next(i int) int {
	return (i + 1) % len(b.data)
}

func (b *Buffer) full() bool {
	return b.next(b.write) == b.read
}

func (b *Buffer) empty() bool {
	return b.write == b.read
}

// ReadByte returns the next byte or an error when empty.
func (b *Buffer) ReadByte() (byte, error) {
	if b.empty() {
		return 0, errors.New("Empty")
	}
	data := b.data[b.read]
	b.read = b.next(b.read)
	return data, nil
}

// WriteByte adds a byte to the buffer or returns an error when full.
func (b *Buffer) WriteByte(c byte) error {
	if b.full() {
		return errors.New("Full")
	}
	b.data[b.write] = c
	b.write = b.next(b.write)
	return nil
}

// Overwrite writes a byte. If the buffer is full, the oldest byte is dropped.
func (b *Buffer) Overwrite(c byte) {
	if b.full() {
		b.read = b.next(b.read)
	}
	b.WriteByte(c)
}

// Reset clears the buffer.
func (b *Buffer) Reset() {
	b.read = 0
	b.write = 0
}
